### PR TITLE
#26: Color memory

### PR DESCRIPTION
**Pull request by shubertm**
#26

### CODE REVIEW

File: app/src/main/java/space/taran/arkretouch/presentation/edit/EditViewModel.kt [hunk 1]
```diff
@@ -137,6 +139,40 @@ class EditViewModel(
...
+                editManager.oldColors.forEach { color ->
+                    val line = color.value.toString()
+                    writer.appendLine(line)
+                }
+                writer.close()
+            } catch (e: IOException) {
+                e.printStackTrace()
+            }
+        }
+    }
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arkretouch/presentation/edit/EditViewModel.kt]: Extract it into `data/Preferences`, add to di component and pass into `EditViewModel` constructor
Just add `private val prefernces: Preferences` to `EditViewModelFactory` constructor

Comment (shubertm) [reply app/src/main/java/space/taran/arkretouch/presentation/edit/EditViewModel.kt]: Alright

**Comment by kirillt** [comment]
Now, it silently crashes for me.. Btw do we have ACRA integrated into Retouch? I don't see crash report dialog.

File: app/src/main/java/space/taran/arkretouch/data/Preferences.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            try {
+                val writer = appContext.openFileOutput(
+                    COLORS_STORAGE,
+                    Context.MODE_PRIVATE
+                ).bufferedWriter()
+                editManager.oldColors.forEach { color ->
+                    val line = color.value.toString()
+                    writer.appendLine(line)
+                }
+                writer.close()
```
Comment (kirillt) [inline app/src/main/java/space/taran/arkretouch/data/Preferences.kt]: Closing should be in `finally` block.

@mdrlzy I think there is dedicated pattern in Kotlin for guaranteed resource closing, right?

File: app/src/main/java/space/taran/arkretouch/data/Preferences.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            try {
+                val read = async {
+                    val reader = appContext.openFileInput(COLORS_STORAGE)
+                        .bufferedReader()
+                    editManager.clearOldColors()
+                    reader.readLines().forEach { line ->
+                        val color = Color(line.toULong())
+                        editManager.addColor(color)
+                    }
+                    reader.close()
```
Comment (kirillt) [inline app/src/main/java/space/taran/arkretouch/data/Preferences.kt]: Same about closing here.

**Comment by shubertm** [comment]
> Now, it silently crashes for me.. Btw do we have ACRA integrated into Retouch? I don't see crash report dialog.

No ACRA

**Comment by kirillt** [comment]
```
java.io.FileNotFoundException: /data/user/0/space.taran.arkretouch2/files/colors: open failed: ENOENT (No such file or directory)
        at libcore.io.IoBridge.open(IoBridge.java:575)
        at java.io.FileInputStream.<init>(FileInputStream.java:160)
        at android.app.ContextImpl.openFileInput(ContextImpl.java:706)
        at android.content.ContextWrapper.openFileInput(ContextWrapper.java:224)
        at space.taran.arkretouch.data.Preferences$readUsedColors$2$read$1.invokeSuspend(Preferences.kt:42)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
    	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@ba92dd6, Dispatchers.Main.immediate]
    Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
        at libcore.io.Linux.open(Native Method)
        at libcore.io.ForwardingOs.open(ForwardingOs.java:567)
        at libcore.io.BlockGuardOs.open(BlockGuardOs.java:273)
        at libcore.io.ForwardingOs.open(ForwardingOs.java:567)
        at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:7756)
        at libcore.io.IoBridge.open(IoBridge.java:561)
        	... 12 more
```

File: app/src/main/java/space/taran/arkretouch/data/Preferences.kt (hunk 2 above)
Comment (shubertm) [reply app/src/main/java/space/taran/arkretouch/data/Preferences.kt]: For that we need a ```writer``` reference outside ```try``` ```catch```

**Comment by shubertm** [comment]
Is this the silent crash log or another?

File: app/src/main/java/space/taran/arkretouch/data/Preferences.kt (hunk 2 above)
Comment (kirillt) [reply app/src/main/java/space/taran/arkretouch/data/Preferences.kt]: @ShubertMunthali yes, but it's common pattern in Java:
https://www.baeldung.com/java-finally-keyword

Kotlin has better feature for guaranteed resource closing, called `use`:
https://www.baeldung.com/kotlin/try-with-resources

File: app/src/main/java/space/taran/arkretouch/data/Preferences.kt [hunk 4]
```diff
@@ -17,14 +19,12 @@ class Preferences @Inject constructor() {
...
-                ).bufferedWriter()
-                editManager.oldColors.forEach { color ->
-                    val line = color.value.toString()
-                    writer.appendLine(line)
+                val colorsStorage = appContext.filesDir.resolve(COLORS_STORAGE)
+                    .toPath()
+                val lines = editManager.oldColors.map { color ->
+                    color.value.toString()
                 }
+                Files.write(colorsStorage, lines, UTF_8)
```
Comment (kirillt) [inline app/src/main/java/space/taran/arkretouch/data/Preferences.kt:28]: That also works

File: app/src/main/java/space/taran/arkretouch/data/Preferences.kt (hunk 2 above)
Comment (kirillt) [reply app/src/main/java/space/taran/arkretouch/data/Preferences.kt]: I see you've made it without explicit reader/writer values. That's even better.

**Review by kirillt — APPROVED** [review]
Feels good

File: app/src/main/java/space/taran/arkretouch/presentation/drawing/EditManager.kt [hunk 5]
```diff
@@ -299,7 +291,16 @@ class EditManager {
         drawPaint.value.color = color
         _currentPaintColor.value = color
         if (oldColors.isNotEmpty() && oldColors.last() != color) {
-            _oldColors.add(color)
+            val size = oldColors.size
+            when {
+                size < 20 -> {
```
Comment (kirillt) [inline app/src/main/java/space/taran/arkretouch/presentation/drawing/EditManager.kt]: Extract `20` as named constant please. Better for maintenance.

File: app/src/main/java/space/taran/arkretouch/presentation/drawing/EditManager.kt [hunk 6]
```diff
@@ -299,7 +291,16 @@ class EditManager {
         drawPaint.value.color = color
         _currentPaintColor.value = color
         if (oldColors.isNotEmpty() && oldColors.last() != color) {
-            _oldColors.add(color)
+            val size = oldColors.size
```
Comment (kirillt) [inline app/src/main/java/space/taran/arkretouch/presentation/drawing/EditManager.kt]: Slightly better code for this:
```
_oldColors.add(color)
if (oldColors.size > MAX_COLORS_AMOUNT) {
    _oldColors = _oldColors.takeLast(MAX_COLORS_AMOUNT)
}
```

File: app/src/main/java/space/taran/arkretouch/presentation/drawing/EditManager.kt [hunk 7]
```diff
@@ -291,16 +291,9 @@ class EditManager {
...
-                    _oldColors.add(color)
-                }
-                size == 20 -> {
-                    _oldColors.removeFirst()
-                    _oldColors.add(color)
-                }
-            }
+            _oldColors.add(color)
+            if (oldColors.size > MAX_COLOR_SIZE)
+                _oldColors.removeFirst()
```
Comment (kirillt) [inline app/src/main/java/space/taran/arkretouch/presentation/drawing/EditManager.kt:296]: I can imagine really rare corner case when `_oldColors = _oldColors.takeLast(MAX_COLORS_AMOUNT)` will be better, but ok ;-)

Comment (shubertm) [reply app/src/main/java/space/taran/arkretouch/presentation/drawing/EditManager.kt:296]: Yes, what is it?

**Pull request merged**